smt/rem_u_i64.py

```diff
@@ -0,0 +1,51 @@
+from z3 import *
+from utils import *
+from functools import reduce
+
+s = init_z3_solver()
+
+# define spec
+lhs, rhs, res_m = Ints('lhs rhs res_m')
+
+wasm_rem_u_i64 = Function('WasmRemUI64', IntSort(), IntSort(), IntSort())
+s.add(ForAll([lhs, rhs], wasm_rem_u_i64(lhs, rhs) == lhs % rhs))
+
+# define var
+aux1, aux2, aux3 = Ints('aux1, aux2, aux3')
+# give name to intermediate expression to make z3 faster
+intermediate = Int('intermediate')
+
+constrain = Function('Constrain', IntSort(), IntSort(), IntSort(
+), IntSort(), IntSort(), IntSort(), IntSort(), BoolSort())
+constraints = [
+    is_i64(aux1),
+    is_i64(aux2),
+    is_i64(aux3),
+    is_i64(lhs),
+    is_i64(rhs),
+    is_i64(res_m),
+    # c.bin.div_u/rem_u
+    fr_add(fr_mul(rhs, aux1), aux2) == intermediate,
+    fr_sub(intermediate, lhs) == 0,
+    fr_sub(fr_add(fr_add(aux2, aux3), 1), rhs) == 0,
+    fr_sub(res_m, aux2) == 0,
+]
+
+s.add(ForAll([aux1, aux2, aux3, intermediate, lhs, rhs, res_m], And(constrain(
+    aux1, aux2, aux3, intermediate, lhs, rhs, res_m) == reduce(lambda x, y: And(x, y), constraints))))
+
+s.push()
+# Completeness
+s.add(And(wasm_rem_u_i64(lhs, rhs) == res_m, ForAll([aux1, aux2, aux3, intermediate, lhs, rhs, res_m], Not(
```

Comment:
1. wasm_rem_u_i64 doesn't constrains that lhs and rhs are i64
2. the lhs and rhs should not in ForAll

Reply:
What's the scope of lhs and rhs for wasm_rem_u_i64? I found they are u64 in https://github.com/DelphinusLab/zkWasm/blob/f9a145fbf52ac0a7f2faed72b7a35baecc2b8109/src/circuits/etable/op_configure/op_bin.rs#L176-L184

Reply:
They are u64, but this line doesn't declare it.
I suppose this line should be
```
Forall lhs, rhs, res_m,
lhs is u64 /\ rhs is u64 ->
(wasm_rem_u_i64(lhs, rhs) == res_m /\ (Forall aux1, aux2, aux3, intermediate, not constrain(aux1, aux2, aux3, intermediate, lhs, rhs, res_m))
```
because we want to prove that
```
Forall lhs, rhs, res_m,
lhs is u64 /\ rhs is u64 ->
wasm_rem_u_i64(lhs, rhs) == res_m ->
(Exists aux1, aux2, aux3, intermediate, constrain(aux1, aux2, aux3, intermediate, lhs, rhs, res_m)
```

Reply:
Updated. Now it proves the following is unsatisfied.
```
Exists lhs, rhs,
lhs is u64 /\ rhs is u64 /\ rhs != 0 ->
wasm_div_u_i64(lhs, rhs) == res_d /\wasm_rem_u_i64(lhs, rhs) == res_m /\ Forall aux1, aux2, aux3, intermediate, not constrain(aux1, aux2, aux3, intermediate, lhs, rhs, res_d, res_m))
```